File: MediaData/src/main/java/com/xu/mediadata/utils/ArtistLoader.java
package com.xu.mediadatatest.utils;

import android.content.Context;
import android.provider.MediaStore;
import android.text.TextUtils;

import com.xu.mediadatatest.entity.Artist;
import com.xu.mediadatatest.entity.Music;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * desc:
 * *
 * user: xujj
 * time: 2022/8/26 14:45
 **/
public class ArtistLoader {

    /**
     * 通过歌曲列表获取歌手列表
     *
     * @param musics //按歌手排序的歌曲列表
     */
    private static List<Artist> getArtistList(List<Music> musics, String name) {
        List<Artist> artistList = new ArrayList<>();
        Map<Long, Artist> artistMap = new LinkedHashMap<>();
        try {
            for (Music music : musics) {
                long artistId = music.getArtistId();
                String artistName = music.getArtist();

                Artist artist = artistMap.get(artistId);
                if (artist == null) {
                    artist = new Artist();
                    artist.setId(artistId);
                    artist.setName(artistName);
                    artist.setMusicCount(1);
                    artistMap.put(artistId, artist);
                } else {
                    artist.setMusicCount(artist.getMusicCount() + 1);
                    artistMap.put(artistId, artist);
                }
            }
            Set<Long> set = artistMap.keySet();
            for (Long artistId : set) {
                Artist artist = artistMap.get(artistId);
                if (artist != null) {
                    if (TextUtils.isEmpty(name)) {
                        artistList.add(artist);
                    } else {
                        if (artist.getName().toLowerCase().contains(name.toLowerCase())) {
                            artistList.add(artist);
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return artistList;
    }

    public static List<Artist> getArtistList(Context context, String name) {
        List<Music> musics = MusicLoader.getMusicList(context, null, MediaStore.Audio.Media.ARTIST_KEY, null);
        return getArtistList(musics, name);
    }

    public static List<Artist> getArtistList(Context context) {
        return getArtistList(context, null);
    }
}

File: MediaData/src/main/java/com/tool/mediadata/utils/ArtistLoader.kt
package com.tool.mediadata.utils

import android.content.Context
import com.tool.mediadata.MediaConfig
import com.tool.mediadata.entity.Artist
import com.tool.mediadata.entity.Music
import com.tool.mediadata.utils.MusicLoader.sorted

/**
 * desc:
 **
 * user: xujj
 * time: 2023/10/20 16:58
 **/
object ArtistLoader {

    fun getArtistListBase(
        musics: List<Music>?,
        name: String? = null,
        sortOrder: String? = MediaConfig.getInstance().artistSortOrder
    ): MutableList<Artist> {
        if (musics.isNullOrEmpty()) return ArrayList()
        val musicList = musics.sorted(SortOrder._ID)

        val artistList = ArrayList<Artist>()
        val artistMap = LinkedHashMap<Long, Artist>()
        try {
            for (music in musicList) {
                val artistId = music.artistId
                val artistName = music.artist
                var artist: Artist? = artistMap.get(artistId)
                if (artist == null) {
                    artist = Artist(artistId, artistName, 1)
                    artistMap.put(artistId, artist)
                } else {
                    artist.musicCount = artist.musicCount + 1
                    artistMap.put(artistId, artist)
                }
            }
            val set: Set<Long> = artistMap.keys
            for (artistId in set) {
                val artist: Artist? = artistMap.get(artistId)
                if (artist != null) {
                    if (name.isNullOrEmpty()) {
                        artistList.add(artist)
                    } else {
                        if (artist.name.lowercase().contains(name.lowercase())) {
                            artistList.add(artist)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        when (sortOrder) {
            SortOrder.ArtistSortOrder.ARTIST_A_Z -> {
                artistList.sortBy {
                    it.name
                }
            }

            SortOrder.ArtistSortOrder.ARTIST_Z_A -> {
                artistList.sortByDescending {
                    it.name
                }
            }

            SortOrder.ArtistSortOrder.ARTIST_MUSIC_COUNT -> {
                artistList.sortBy {
                    it.musicCount
                }
            }

            SortOrder.ArtistSortOrder.ARTIST_MUSIC_COUNT_DESC -> {
                artistList.sortByDescending {
                    it.musicCount
                }
            }
        }
        return artistList
    }

    fun getArtistList(
        context: Context?,
        name: String? = null,
        sortOrder: String? = MediaConfig.getInstance().artistSortOrder
    ): MutableList<Artist> {
        val musicList = MusicLoader.getAllMusicList(context)
        return getArtistListBase(musicList, name, sortOrder)
    }
}